### ISSUE TITLE
[nextcloud] latest image stops after "unhandled exception"

### ISSUE THREAD

**Issue by nrgeen** [Wonderfall/dockerfiles#130]
Good Morning,
Since yesterday evening for me the latest image build isn't working any more.
In the log I get the following message. In the next step the container stops every time.

```
Updating permissions...
Updating permissions in /nextcloud...
Permissions in /data are correct.
Permissions in /config are correct.
Permissions in /apps2 are correct.
Updating permissions in /etc/nginx...
Updating permissions in /etc/php7.1...
Updating permissions in /var/log...
Updating permissions in /var/lib/nginx...
Updating permissions in /tmp...
Updating permissions in /etc/s6.d...
Done updating permissions.
An unhandled exception has been thrown:
Exception: Could not gather sufficient random data in /nextcloud/lib/private/Security/SecureRandom.php:81

ace:
#0 /nextcloud/lib/private/Security/SecureRandom.php(81): random_int(0, 63)
#1 /nextcloud/lib/private/Security/CSRF/CsrfTokenGenerator.php(51): OC\Security\SecureRandom->generate(32)
#2 /nextcloud/lib/private/Security/CSRF/CsrfTokenManager.php(63): OC\Security\CSRF\CsrfTokenGenerator->generateToken()
#3 /nextcloud/lib/private/Security/CSP/ContentSecurityPolicyNonceManager.php(56): OC\Security\CSRF\CsrfTokenManager->getToken()
#4 /nextcloud/lib/private/legacy/response.php(250): OC\Security\CSP\ContentSecurityPolicyNonceManager->getNonce()
#5 /nextcloud/lib/base.php(684): OC_Response::addSecurityHeaders()
#6 /nextcloud/lib/base.php(1092): OC::init()
#7 /nextcloud/console.php(48): require_once('/nextcloud/lib/...')
#8 /nextcloud/occ(11): require_once('/nextcloud/cons...')
#9 {main}Trying ownCloud upgrade again to work around ownCloud upgrade bug...
An unhandled exception has been thrown:
Exception: Could not gather sufficient random data in /nextcloud/lib/private/Security/SecureRandom.php:81

ace:
#0 /nextcloud/lib/private/Security/SecureRandom.php(81): random_int(0, 63)
#1 /nextcloud/lib/private/Security/CSRF/CsrfTokenGenerator.php(51): OC\Security\SecureRandom->generate(32)
#2 /nextcloud/lib/private/Security/CSRF/CsrfTokenManager.php(63): OC\Security\CSRF\CsrfTokenGenerator->generateToken()
#3 /nextcloud/lib/private/Security/CSP/ContentSecurityPolicyNonceManager.php(56): OC\Security\CSRF\CsrfTokenManager->getToken()
#4 /nextcloud/lib/private/legacy/response.php(250): OC\Security\CSP\ContentSecurityPolicyNonceManager->getNonce()
#5 /nextcloud/lib/base.php(684): OC_Response::addSecurityHeaders()
#6 /nextcloud/lib/base.php(1092): OC::init()
#7 /nextcloud/console.php(48): require_once('/nextcloud/lib/...')
#8 /nextcloud/occ(11): require_once('/nextcloud/cons...')

}

```

The container is also not working without any links to other containers and without any volume-mapping. The same error occure.

Thanks for your help!!!

**Comment by Wonderfall** [comment]
I can't reproduce... What does `docker info` return?

**Comment by nrgeen** [comment]
Thanks for your fast reply! 

**docker info:**
```
Containers: 8
 Running: 8
 Paused: 0
 Stopped: 0
Images: 9
Server Version: 1.11.2
Storage Driver: btrfs
Logging Driver: db
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host bridge null
Kernel Version: 3.10.102
Operating System: <unknown>
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.873 GiB
Name: Diskstation
ID: XXX
Docker Root Dir: /volume1/@docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No kernel memory limit support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
```

**Comment by Wonderfall** [comment]
Just to be sure before digging into details : 
- Are you using the latest image with PHP 7.1.2?
- Does it fail like this when running `docker run -ti --rm wonderfall/nextcloud:11.0`?

Otherwise at first sight I could say this is related to the kernel (common issue) but if it was working before, it is weirder. https://github.com/owncloud/core/issues/22983

Also it could be related to : https://github.com/docker-library/php/issues/376
And the fact you're using a "relatively old" kernel version (3.10) confort me in my opinion. I can't reproduce because I'm currently using Linux 4.9.x. 

In other words, a fix is directly coming to the Alpine PHP packages, you have to wait and I'll launch a new automated build when it's available.

Or you can upgrade your kernel, I recommend at least 3.17 because this is when getrandom() was introduced. :)

**Comment by ghost** [comment]
Same here! I just pulled the latest docker image too.
And and similarly I have a relatively old kernel: `Linux 3.16.0-4-amd64`

**Comment by Wonderfall** [comment]
Immediately the only solution is to upgrade you kernel to at least 3.17. `php7` packages are also affected with `php7.1` so there's nothing else at the moment.

**Comment by hoellen** [comment]
Same here: `Linux 3.10.0`

**Comment by SemaiCZE** [comment]
I need it to work on my OpenVZ VPS where is an old kernel by design without any upgrade possibility. I'm using Docker 1.10 version (the latest working version with OpenVZ) on CentOS 7. I'll keep older build for now.

**Comment by Wonderfall** [comment]
Not really up to me. :/
However to make things faster I've juste flagged php 7.1 packages and asked the maintainer if it's possible to backport this fix and make a r1 version.

**Comment by xataz** [comment]
Hi,

Same problem after upgrade container today.
Debian 8 with 3.16 kernel
Docker 1.13 (btrfs)

`docker info` : 
```shell
Containers: 18
 Running: 16
 Paused: 0
 Stopped: 2
Images: 12
Server Version: 1.13.0
Storage Driver: btrfs
 Build Version: Btrfs v3.17
 Library Version: 101
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e
runc version: 2f7393a47307a16f8cee44a37b262e8b81021e3e
init version: 949e6fa
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 31.46 GiB
Name: xxxxxx
ID: W6BC:O6VC:P62D:NK67:2E2M:3VTM:6QBX:X6NC:TYMI:DSF4:FJOX:USEK
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No kernel memory limit support
WARNING: No oom kill disable support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by Wonderfall** [comment]
I know I know, all kernels under 3.17 are affected!

**Comment by nrgeen** [comment]
Oh wow,  it's a hive of activity here.

For me it is now time to update my hardware. My NAS will now return to its original function. ;-)

So thank you very much for your support and especially your commitment!!!!

**Comment by vakartel** [comment]
@Wonderfall 
Just post backported patches for php7 & php7.1 to alpinelinux patchwork.
http://patchwork.alpinelinux.org/project/aports/list/?q=74105&submitter=&archive=&delegate=&state=*
so wait when it would be accepted

**Comment by xataz** [comment]
Ok with 4.9 kernel.

**Comment by Wonderfall** [comment]
@vakartel Thanks for your reactivity! 

I'll trigger an automated build once it's accepted then. :)

**Comment by vakartel** [comment]
@Wonderfall 
If you want to try it before it would be accepted, you can add my local unofficial repo just for testing (x86_64 only) http://alpinelinux.org.ua/vaka/

add to /etc/apk/repositories:
   http://alpinelinux.org.ua/vaka/main
   http://alpinelinux.org.ua/vaka/testing
   http://alpinelinux.org.ua/vaka/community

public key you need to copy to /etc/apk/keys/ to use this repo:
http://alpinelinux.org.ua/vaka/valery.kartel%40gmail.com-5476f68b.rsa.pub

There is a stuff not accepted to official yet, so do not system-wide/docker-wide 'apk upgrade' with this enabled (I have my own php5 and openssh changes there)

**Comment by Wonderfall** [comment]
@vakartel Nice idea, I've just made a new tag : `11.0-fix` using your php-7.1.2 r1 packages. I think it's still safe to do a system-wide upgrade (it is necessary for php-gd to work because the base image was not updated for a while) while I'm using the `testing` tag (for packages).

I invite those who have this issue to try this image. :)

EDIT : It's [currently building...](https://hub.docker.com/r/wonderfall/nextcloud/builds/)

**Comment by skurtzemann** [comment]
@Wonderfall 
I try the new image `11.0-fix` with a 3.10.0 kernel and it works ! Thanks

**Comment by cragr** [comment]
@skurtzemann 
I also confirm a 3.10.0 kernel working with the 11.0-fix tag.

**Comment by vakartel** [comment]
@Wonderfall 
patches accepted into official repo ;)

**Comment by Wonderfall** [comment]
Thanks again.

**Comment by vakartel** [comment]
@Wonderfall 

-- sed -i "$ s|\-n||g" /usr/bin/pecl

it fixed in php7.1, so this hack no longer needed :)

PS: php5 & php7 not fixed yet

**Comment by macmeck** [comment]
@vakartel any update on this for PHP 7.0? we're seeing this issue again since @Wonderfall had to go back from edge (which has PHP7.1) to alpine 3.5 (which still has PHP 7.0) :-|

**Comment by Steiniche** [comment]
I am seeing this issue as well.
Unfortunately Synology NAS' have some very old kernels: 3.10.102
Can anyone tell me what the problem was running on alpine:edge?

**Comment by klaxa** [comment]
@Steiniche nginx package cannot be installed. See: https://bugs.alpinelinux.org/issues/7187

**Comment by Steiniche** [comment]
@klaxa thanks for pointing that out to me.
Got the same error when I tried to build the Dockerfile with alpine:edge.
So all I/we can do is wait for the alpine guys to fix the libressl issue?

**Comment by klaxa** [comment]
@Steiniche Pretty much... @Wonderfall wanted to build an nginx-php container that builds both from source and I think there are already commits for it, but the nextcloud image isn't based off of that yet.

**Comment by Steiniche** [comment]
@klaxa Well it seems that we are in luck. 
The issue is being prioritized by the libressl guys: https://github.com/libressl-portable/portable/issues/307
